url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/edu/luc/nmerge/mvd/MVDFile.java

```diff
@@ -0,0 +1,771 @@
+ */
+
+package edu.luc.nmerge.mvd;
+import java.sql.*;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.File;
+import java.io.UnsupportedEncodingException;
+import java.util.BitSet;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.ListIterator;
+import java.util.Properties;
+import edu.luc.nmerge.exception.*;
+
+/**
+ *	Load and save an MVD file in binary form. The raw MVD binary 
+ *	structure is first ZIP encoded and then Base64 wrapped.
+ */
+public class MVDFile 
+{
+	public static boolean debug = false;
+	/** magic string '0xC0DEDEAD' */
+	static byte[] MVD_MAGIC = {(byte)'\336',(byte)'\255',
+		(byte)'\300',(byte)'\336'};
+	/**
+	 * Write to a Mysql database named mvd which has a table called works 
+	 * with fields name, description and data.
+	 * @param file the name of the file (key of table)
+	 * @param description the description of the mvd
+	 * @param data a String with the MVD data encoded in base64
+	 * @param folderId id of the containing folder
+	 * @param properties property file containing db connection
+	 * @throws Exception
+	 */
+	private static void writeToDatabase( String file, String data, 
+		String description, int folderId, Properties props ) throws Exception
+	{
+		Connection conn = null;
+		ResultSet rs = null;
+		Statement stmt = null;
+		String username = (String)props.get("username");
+		String password = (String)props.get("password");
+		String url = (String)props.get("jdbc-url");
+		String jdbcClass = (String)props.get("jdbc-class");
+		String dbName = (String)props.get("mvd-db-name");
+		String nameField = (String)props.get("mvd-name-field");
+		String descField = (String)props.get("mvd-desc-field");
+		String dataField = (String)props.get("mvd-data-field");
+		String tableName = (String)props.get("mvd-table-name");
+		String folderField = (String)props.get("mvd-folder-field");
+/*		// debug
+		Set<Object> keys = props.keySet();
+		Iterator<Object> iter = keys.iterator();
+		while ( iter.hasNext() )
+		{
+			String key = (String) iter.next();
+			String value = (String)props.getProperty(key);
+		}
+		// end debug
+ */
+		String name = (file.endsWith(".mvd"))
+			?file.substring(0,file.length()-4)
+			:file;
+		description = description.replace("\"", "\\\"");
+		if ( username == null )
+			username = "root";
+		if ( password == null )
+			password = "jabberw0cky";
+		if ( url == null )
+			url = "jdbc:mysql://localhost:3306/";
+		if ( jdbcClass == null )
+			jdbcClass = "com.mysql.jdbc.Driver";
+		if ( dbName == null )
+			dbName = "dv";
+		if ( nameField == null )
+			nameField = "name";
+		if ( descField == null )
+			descField = "description";
+		if ( dataField == null )
+			dataField = "data";
+		if ( tableName == null )
+			tableName = "works";
+		if ( folderField == null )
+			folderField = "folder_id";
+		try
+        {
+            Class.forName(jdbcClass).newInstance();
+            conn = DriverManager.getConnection(url+dbName, username, password);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *conn*. The resource is referenced at lines 92, 143, 145, 165, 173, 190. The resource is closed at lines 168, 195. There are other execution paths that do not contain closure statements, for example, when Connection.createStatement throws an exception. Ensure *conn* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).